### ISSUE TITLE
Error in Moodle "3.7.1+": Call to undefined method MoodleQuickForm_static::getElements()

### ISSUE THREAD

**Issue by dionysius** [frankkoch/moodle-mod_studentquiz#159]
- Error happens on current moodle branch `MOODLE_37_STABLE`, identified as version `3.7.1+`, sitting on commit https://github.com/moodle/moodle/commit/684d142401f82761064367d027b11ff701be95cc
- **Does not happen** on tag `v3.7.1`, so it's in between.

```bash
$ git checkout MOODLE_37_STABLE 
Previous HEAD position was b198428e41b Moodle release 3.7.1
Switched to branch 'MOODLE_37_STABLE'
Your branch is up to date with 'origin/MOODLE_37_STABLE'.

$ git bisect start

$ git bisect bad

$ git bisect good v3.7.1
Bisecting: 111 revisions left to test after this (roughly 7 steps)
[b0d81fce7bdfa42de15946ba65cda0eee8a3c0af] Merge branch 'MDL-65651-37' of https://github.com/lucaboesch/moodle into MOODLE_37_STABLE

$ git bisect good
Bisecting: 55 revisions left to test after this (roughly 6 steps)
[4f41f8407d0cf62d964ca766756e3f4855767c5b] Merge branch 'MDL-65561-37' of git://github.com/bmbrands/moodle into MOODLE_37_STABLE

$ git bisect bad
Bisecting: 26 revisions left to test after this (roughly 5 steps)
[03667bad2016e76d45d812c2766e1c3a3742fafb] Merge branch 'install_37_STABLE' of https://git.in.moodle.com/amosbot/moodle-install into MOODLE_37_STABLE

$ git bisect good
Bisecting: 13 revisions left to test after this (roughly 4 steps)
[1893b39bfc5a7536e0dadd9667f2e5fc488f14e3] Merge branch 'm37_MDL-66140_Mitigate_PECL_Solr_issue_Ref_72740' of https://github.com/scara/moodle into MOODLE_37_STABLE

$ git bisect good
Bisecting: 6 revisions left to test after this (roughly 3 steps)
[a408e168cc0a2ec2e847f5889c085cac0d3faeb4] Merge branch 'install_37_STABLE' of https://git.in.moodle.com/amosbot/moodle-install into MOODLE_37_STABLE

$ git bisect good
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[dda635074d5162c2955e192a61eae1257ba1821c] Merge branch 'MDL-65886-MOODLE-37' of https://github.com/kristian-94/moodle into MOODLE_37_STABLE

$ git bisect good
Bisecting: 1 revision left to test after this (roughly 1 step)
[827473833d97ca3228ae310208ead0924fb9bef0] Merge branch 'MDL-65604-37' of git://github.com/andrewnicols/moodle into MOODLE_37_STABLE

$ git bisect good
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[d70fb8fefe30b3211aeca68004212c918ac5aa67] MDL-65561 themes: user date filters display fixes

$ git bisect bad
d70fb8fefe30b3211aeca68004212c918ac5aa67 is the first bad commit
commit d70fb8fefe30b3211aeca68004212c918ac5aa67
Author: Bas Brands <bas@moodle.com>
Date:   Tue Aug 6 13:56:17 2019 +0200

    MDL-65561 themes: user date filters display fixes

:040000 040000 3ddfdf3c0c98f119a9d60db140276df2d5b0918a b69a100f034abc4bf60425d8dcafdbf93dd58613 M      user
```

Seems like this commit https://github.com/moodle/moodle/commit/d70fb8fefe30b3211aeca68004212c918ac5aa67 introduced this error.

![Screenshot from 2019-08-28 17 47 04](https://user-images.githubusercontent.com/1341084/63871114-8ccfb280-c9bb-11e9-9d0b-03183292b158.png)

**Comment by dionysius** [comment]
```diff
diff --git a/classes/question/bank/question_bank_filter.php b/classes/question/bank/question_bank_filter.php
index a29cc68..0f75358 100644
--- a/classes/question/bank/question_bank_filter.php
+++ b/classes/question/bank/question_bank_filter.php
@@ -157,6 +157,7 @@ class studentquiz_user_filter_date extends user_filter_date {
         if (!empty($group) && !($group instanceof HTML_QuickForm_Error)) {
             $groupelements = $group->getElements();
             if (!empty($groupelements)) {
+                var_dump(array_map(function ($a) { return get_class($a);}, $groupelements));
                 $dateselector = $groupelements[1]->getElements();
                 if (!empty($dateselector)) {
                     $isbefore = optional_param('timecreated_sdt', 0, PARAM_INT);
```

`MOODLE_37_STABLE`
```
 array(8) {
  [0]=> string(22) "MoodleQuickForm_static"
  [1]=> string(22) "MoodleQuickForm_static"
  [2]=> string(29) "MoodleQuickForm_date_selector"
  [3]=> string(22) "MoodleQuickForm_static"
  [4]=> string(22) "MoodleQuickForm_static"
  [5]=> string(22) "MoodleQuickForm_static"
  [6]=> string(29) "MoodleQuickForm_date_selector"
  [7]=> string(22) "MoodleQuickForm_static"
}
```

'v3.7.1`
```
array(5) {
  [0]=> string(22) "MoodleQuickForm_static"
  [1]=> string(29) "MoodleQuickForm_date_selector"
  [2]=> string(22) "MoodleQuickForm_static"
  [3]=> string(22) "MoodleQuickForm_static"
  [4]=> string(29) "MoodleQuickForm_date_selector"
}
```

Yeah, hardcoding to select index 1 might not be a good idea...

**Comment by dionysius** [comment]
@HuongNV13 I'm not quite sure what this section exactly does, you've introduced it in https://github.com/frankkoch/moodle-mod_studentquiz/commit/34fcbd80c0286ef126891f355981b91ca69f5541, but it looks like some label and attribute settings for screen readability.

As pointed above, in newest moodle the element list has changed, so we should get rid of the magic constand `1` as the index. What do you think about a for loop until `$el instanceof MoodleQuickForm_date_selector`? Can you review https://github.com/frankkoch/moodle-mod_studentquiz/commit/0f1557e6cc05b4b598fad2d5b4a3dc15c1e3d81b please?

**Comment by steveorulez** [comment]
I have the same error on moodle 3.6.5+

**Comment by dionysius** [comment]
I've released v4.0.4 at least fixing the error with my suggested changes above. Waiting for @HuongNV13 if it can stay that way

**Comment by HuongNV13** [comment]
Hi @dionysius,
The https://github.com/frankkoch/moodle-mod_studentquiz/compare/92e10f77a79b18328102fed56a2b41956b6c1488...9959f089977e4dd82c74a27103612e5b71e3e06f look good

You can use **$el->getType() == 'date_selector'** instead of **$el instanceof MoodleQuickForm_date_selector** to reduce performance

Thanks,
